be/src/exec/pipeline/exchange/sink_buffer.h

```diff
@@ -98,74 +95,28 @@ class SinkBuffer {
+        {
+            std::lock_guard<std::mutex> l(_mutex);
+            if (_received_all_eos) {
+                request.params->release_finst_id();
+                return;
+            }
         }
-        std::lock_guard<std::mutex> l(_mutex);
-        _buffers[request.fragment_instance_id].push(request);
-        _buffer_empty_cv.notify_one();
-    }
-
-    void process() {
-        try {
-            MemTracker* prev_tracker = tls_thread_status.set_mem_tracker(_mem_tracker);
-            DeferOp op([&] { tls_thread_status.set_mem_tracker(prev_tracker); });
-
-            while (!_is_finished) {
-                {
-                    std::unique_lock<std::mutex> l(_mutex);
-                    bool is_buffer_empty = true;
-                    for (auto& [_, buffer] : _buffers) {
-                        if (!buffer.empty()) {
-                            is_buffer_empty = false;
-                            break;
-                        }
-                    }
-                    if (is_buffer_empty) {
-                        _buffer_empty_cv.wait(l);
-                    }
-                }
-
-                const size_t spin_threshould = 100;
-                size_t spin_iter = 0;
 
-                for (; spin_iter < spin_threshould; ++spin_iter) {
-                    bool find_any = false;
-                    for (auto& [_, buffer] : _buffers) {
-                        if (buffer.empty()) {
-                            continue;
-                        }
-
-                        // std::queue' read is concurrent safe without mutex
-                        if (!_closures[buffer.front().fragment_instance_id]->has_in_flight_rpc()) {
-                            TransmitChunkInfo info = buffer.front();
-                            find_any = true;
-                            _send_rpc(info);
-                            {
-                                std::lock_guard<std::mutex> l(_mutex);
-                                buffer.pop();
-                            }
-                            info.params->release_finst_id();
-                        }
-                    }
+        {
+            std::lock_guard<std::mutex> l(_buffer_mutex);
+            _buffers[request.fragment_instance_id].push(request);
+        }
 
-                    if (find_any) {
-                        spin_iter = 0;
-                    }
-                }
+        _try_to_trigger_io_task();
 
-                // Find none ready closure after multiply spin, just wait for a while
-#ifdef __x86_64__
-                _mm_pause();
-#else
-                sched_yield();
-#endif
+        if (request.params->eos()) {
+            _received_eos_num.fetch_add(1, std::memory_order_release);
+            if (_received_eos_num.load(std::memory_order_acquire) == _expected_eos_num) {
+                std::lock_guard<std::mutex> l(_mutex);
+                _received_all_eos = true;
```

Comment:
We could merge `_expected_eos_num`, `_received_eos_num`, `_received_all_eos` to one variable `_num_running_sinkers`.
- Init it  with current `_expected_eos_num`.
- Judge `_received_all_eos` by `_expected_eos_num==0`.

Reply:
good suggestion